### ISSUE TITLE
Sony phones crash when removing badge permission

### ISSUE THREAD

**Issue by iForests** [OneSignal/OneSignal-Android-SDK#129]
I follow the instruction here: [Disable Badges](https://documentation.onesignal.com/docs/android-native-sdk#section-disable-badges), but a lot of Sony phones crash. The error message is:

```
Fatal Exception: java.lang.SecurityException: Missing permission to insert badges
       at android.os.Parcel.readException(Parcel.java:1627)
       at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:183)
       at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:135)
       at android.content.ContentProviderProxy.insert(ContentProviderNative.java:476)
       at android.content.ContentResolver.insert(ContentResolver.java:1240)
       at android.content.AsyncQueryHandler$WorkerHandler.handleMessage(AsyncQueryHandler.java:96)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:224)
       at android.os.HandlerThread.run(HandlerThread.java:61)
```

The affected phones are all Sony's.

Please let me know if I can provide any additional information. Thanks.

**Comment by DimaKoz** [comment]
I got the same issue.

**Comment by iForests** [comment]
Do we need to check the value of `<meta-data android:name="com.onesignal.BadgeCount" android:value="DISABLE" />` before using [BadgeCountUpdater.updateCount()](https://github.com/OneSignal/OneSignal-Android-SDK/blob/master/OneSignalSDK/onesignal/src/main/java/com/onesignal/BadgeCountUpdater.java#L71) for example in [OneSignal.init()](https://github.com/OneSignal/OneSignal-Android-SDK/blob/master/OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java#L290)?

**Comment by jkasten2** [comment]
@iForests You are correct, there are 2 calls being made to `updateCount` directly from `OneSignal.java` which are omitting this manifest check. I'll add a fix for this in our next release.

**Comment by jkasten2** [comment]
@iForests @DimaKoz Fixed this in commit e4e44324c3efc06b3e8e0b1eb09013348ec2d0a0. It will be include in the next minor release.

**Comment by rampr** [comment]
Using the following:
```ShortcutBadger.applyCount(activity, 0);```
and if the permissions are disabled, the crash reappears.

Is this the right call to explicitly clear the badge count?

**Comment by jkasten2** [comment]
@rampr OneSignal 3.4.2 wraps `ShortcutBadger.applyCount` with try catches internally to be safe. If you are using the library ShortcutBadger directly I recommend opening an issue with the repo below.
https://github.com/leolin310148/ShortcutBadger/issues

Note OneSignal 3.4.2 uses version 1.1.11 of ShortcutBadger.

**Comment by rampr** [comment]
@jkasten2. I'm using the ShortcutBadger imported into Onesignal like here -
https://github.com/hashcube/onesignal/blob/master/android/OnesignalPlugin.java#L73

Let me know if we also have to put a try catch around the call or is there a different helper function from onesignal that we can use to clear the count

**Comment by stepan23** [comment]
I have a same problem for Sony

**Comment by jkasten2** [comment]
@stepan23 Are you using the latest OneSignal 3.4.3 SDK? Can you post the fullstack trace from your logcat?

**Comment by stepan23** [comment]
Yes, I do.  I use 3.4.3. I don't have a logcat, because I don't have sony devices. But I have a exception log from google play developer console https://drive.google.com/open?id=0B4yXxYRqlcFnOG9fZEFDcUR3cGc

**Comment by jkasten2** [comment]
@stepan23 I don't see OneSignal in the stacktrace, although I don't see any non Android classes in their either. Seems to be the same as the OP's stacktrace.

Can you insure you have the following in have the following under your application tag in your `AndroidManifest.xml`?
```XML
<meta-data android:name="com.onesignal.BadgeCount" android:value="DISABLE" />
```

Do you get crashes or badges set when they shouldn't from any other launcher?

Do you have any other SDKs / Plugins that could also be setting badges?

If there is an error reading the `AndroidManifest.xml` we default badges on. I can change this to default off but due to the number of crashes you are seeing I don't think this is the issue.
https://github.com/OneSignal/OneSignal-Android-SDK/blob/3.4.3/OneSignalSDK/onesignal/src/main/java/com/onesignal/BadgeCountUpdater.java#L54

**Comment by stepan23** [comment]
1) I have this meta-data in my manifest
2) We began to use your SDK, but our users began to complain about the bandles. We turned them off and now some users have to crash the application.
3) No, we don't have others libs for badges

**Comment by jkasten2** [comment]
@iForests @DimaKoz @rampr @stepan23 I have changed the error handling to disable badges if their is any kind of exception thrown when `AndroidManifest.xml` is read in commit 6ab0a9bbf059477b3fe5a88c08f2d2049414c394.

I have also updated the Shortcut badger library included with this SDK to 1.1.16 in commit aad2e82fce2db728b5408e974e2ee654b9fa8591. It includes a fix for Sony launcher that may have created issues keeping the badge count in sync when badges are not disabled.

These fixes will be included in the 3.4.4 release. Feel free to reopen this issue if the problem persists after updating.
